tests/array.egg

```diff
@@ -20,21 +20,21 @@
 (relation neq (Math Math))
 
 (rule ((neq x y))
-      ((neq y x)))
+      ((set (neq y x) ())))
```

Comment:
Oh I see, now relations have to be set. My opinion is that we shouldn't make users use `set` for relations.

I think the long-term plan is for us to desugar relations to be constructors so people don't have to do this. You could either implement that plan in this PR, or desugar relations to a `set` for now and do that in another PR.

Reply:
I don't think I'll implement the desugaring in this PR. My impression is that it is still up to discussion what it should be desugared into. Performance overhead aside, desugaring into a constructor might not be ideal because that makes it more expressive than needed. For instance, now you can union two edges in the relation or have a relation as the codomain of a function.

Reply:
That's a good point- we would have to disallow unioning two relation terms.
You can't have a relation as the codomain of a function because the sort would be fresh (no way to get a handle on it)

Relation desugaring aside, I don't think we want to make this breaking change when we plan to fix it right away.

Reply:
Let's make functions whose output is `Unit` a special case: Such functions have default value `()` and an implicit merge function.

Reply:
I don't think that would work. I think the plan is to get rid of default values altogether. Desugared or not.

Reply:
I meant getting rid of the `default` field/keyword, but handle unit specially (similar to how EqSort is handled specially).

Reply:
Conclusions:
- Conceptually: 3 function subtypes
  - Constructor | Relation | Custom
  - Operates differently
- Syntax
  - Keeping the status quo
    - Because there is not enough reason to change it
    - The decision is final for the June vision
  - Adding constructor keyword
    - Only relation if declared as relation
    - Only constructor if declared as constructor
- Semantics/implementation
  - Everything still uses the same function data structure under the hood
  - Only relation and constructor are allowed on RHS
     - It adds the relation/creates the term
  - Union requires both arguments to be the same EqSort
  - Set requires the head to be a custom function